File: Book/C/Fig_5.7_dynamicSchedule.c
#include <stdio.h>
#include <math.h>
#include <stdbool.h>
#include <omp.h>

#define ITER 50000000  // use a smaller value if available memory is small

bool check_prime(int num)
{
   int i;
   for (i = 2; i <= sqrt(num); i++) {
      if (num % i == 0)
         return false;
   }
   return true;
}

void main( )
{
   int sum = 0;

   #pragma omp parallel 
   {
      int i;
      int id = omp_get_thread_num();
      double tdata = omp_get_wtime();
 
      #pragma omp for reduction (+:sum) schedule(dynamic)
      for (i = 2; i <= ITER ; i++) {
         if (check_prime(i)) sum++;
      }
      tdata = omp_get_wtime() - tdata;

      if (id == 0) printf("Number of prime numbers is %d in %f sec \n", sum, tdata);
   }
}

File: Book/Fortran/Fig_5.7_dynamicSchedule.f90
program main
 
   use omp_lib
   implicit none

   integer, parameter :: ITER = 50000000   ! use a smaller value if available memory is small
   integer :: i, id
   real*8 :: tdata
   integer :: sum = 0

   !$omp parallel private (i, id, tdata)
      id = omp_get_thread_num()
      tdata = omp_get_wtime()
 
   !$omp do reduction (+:sum) schedule(dynamic)
      do i = 2, ITER
         if (check_prime(i) == 1) sum = sum + 1 
      enddo
   !$omp end do

      tdata = omp_get_wtime() - tdata

      if (id == 0) print *, "Number of prime numbers is ", &
                   & sum, "in ", tdata, " sec"
   !$omp end parallel

   contains
      integer function check_prime (num) 
         implicit none
         integer, intent (in) :: num
         integer :: i, iend

         iend = int (sqrt(num*1.0))
         do i = 2, iend
            if (mod(num,i) == 0) then
                check_prime = 0
                return
            endif
         enddo

         check_prime = 1

      end function check_prime

end program main



